src/logging/timestamp_checkpoint_manager.cpp

```diff
@@ -0,0 +1,1084 @@
+}
+
+void TimestampCheckpointManager::CheckpointingTableDataWithoutTileGroup(
+    const storage::DataTable *table, const cid_t &begin_cid,
+    FileHandle &file_handle) {
+  CopySerializeOutput output_buffer;
+
+  LOG_TRACE("Do checkpointing without tile group to table %d in database %d",
+            table->GetOid(), table->GetDatabaseOid());
+
+  // load all table data without tile group information
+  size_t tile_group_count = table->GetTileGroupCount();
+  for (oid_t tg_offset = START_OID; tg_offset < tile_group_count; tg_offset++) {
+    auto tile_group = table->GetTileGroup(tg_offset);
+    auto tile_group_header = tile_group->GetHeader();
+
+    // load visible tuples data in the table
+    auto max_tuple_count = tile_group->GetNextTupleSlot();
+    auto column_count = table->GetSchema()->GetColumnCount();
+    for (oid_t tuple_id = START_OID; tuple_id < max_tuple_count; tuple_id++) {
+      if (IsVisible(tile_group_header, tuple_id, begin_cid)) {
+        // load all field data of each column in the tuple
+        for (oid_t column_id = START_OID; column_id < column_count;
+             column_id++) {
+          type::Value value = tile_group->GetValue(tuple_id, column_id);
+          value.SerializeTo(output_buffer);
+          LOG_TRACE("%s(column %d, tuple %d):%s\n", table->GetName().c_str(),
+                    column_id, tuple_id, value.ToString().c_str());
+        }
+      } else {
+        LOG_TRACE("%s's tuple %d is invisible\n", table->GetName().c_str(),
+                  tuple_id);
+      }
+    }
+
+    // write down tuple data to file
+    int ret = fwrite((void *)output_buffer.Data(), output_buffer.Size(), 1,
+                     file_handle.file);
+    if (ret != 1 && ret != 0) {
+      LOG_ERROR("Write error: %d", ret);
+      return;
+    }
+
+    output_buffer.Reset();
+  }
+
+  LoggingUtil::FFlushFsync(file_handle);
+}
+
+bool TimestampCheckpointManager::IsVisible(
+    const storage::TileGroupHeader *header, const oid_t &tuple_id,
+    const cid_t &begin_cid) {
+  txn_id_t tuple_txn_id = header->GetTransactionId(tuple_id);
+  cid_t tuple_begin_cid = header->GetBeginCommitId(tuple_id);
+  cid_t tuple_end_cid = header->GetEndCommitId(tuple_id);
+
+  // the tuple has already been committed
+  bool activated = (begin_cid >= tuple_begin_cid);
+  // the tuple is not visible
+  bool invalidated = (begin_cid >= tuple_end_cid);
+
+  if (tuple_txn_id == INVALID_TXN_ID) {
+    // this tuple is not available.
+    return false;
+  }
+
+  if (tuple_txn_id == INITIAL_TXN_ID) {
+    // this tuple is not owned by any other transaction.
+    if (activated && !invalidated) {
+      return true;
+    } else {
+      return false;
+    }
+  } else {
+    // this tuple is owned by othre transactions.
+    if (tuple_begin_cid == MAX_CID) {
+      // this tuple is an uncommitted version.
+      return false;
+    } else {
+      if (activated && !invalidated) {
+        return true;
+      } else {
+        return false;
+      }
```

Comment:
Same as above. Simplify the if-else statement to `return activated && !invalidated;`